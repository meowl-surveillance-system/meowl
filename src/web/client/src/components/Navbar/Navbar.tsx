import React, { Component } from "react";
import { Link as RouterLink } from "react-router-dom";
import AppBar from "@material-ui/core/AppBar";
import Grid from "@material-ui/core/Grid";
import Toolbar from "@material-ui/core/Toolbar";
import Typography from "@material-ui/core/Typography";

interface Props {
  isLoggedIn: boolean;
  isAdmin: boolean;
  onAuthChange: (authState: boolean, adminState: boolean) => void;
}
interface State { }

/**
 * A navigation bar component for easy access to other pages
 */
export default class Navbar extends Component<Props, State> {
  /**
   * Logout the user by sending a logout request to the auth server
   */
  logoutSubmit = (): void => {
    const requestOptions = {
      method: "POST",
    };
    fetch(`/auth/logout`, requestOptions);
    this.props.onAuthChange(false, false);
  };

  /**
   * Renders the button on navbar when user is admin
   */
  renderIsAdmin = () => {
    return (
      <RouterLink
        to="/pendingAccounts"
        style={{
          paddingLeft: "10px",
          color: "inherit",
          textDecoration: "none",
        }}
      >
        PendingAccounts
      </RouterLink>
    );
  };

  /**
   * Renders the buttons on the navbar when the user is authenticated
   */
  renderIsLoggedIn = () => {
    return (
      <div>
        <RouterLink
          to="/uploadTrainingData"
          style={{ color: "inherit", textDecoration: "none" }}
        >
          UploadTrainingData
        </RouterLink>
        <RouterLink
          to="/liveStreams"
          style={{ paddingLeft: "10px", color: "inherit", textDecoration: "none" }}
        >
          LiveStreams
        </RouterLink>
        <RouterLink
          to="/streams"
          style={{
            paddingLeft: "10px",
            color: "inherit",
            textDecoration: "none",
          }}
        >
          Streams
        </RouterLink>
        {this.props.isAdmin && this.renderIsAdmin()}
        <RouterLink
          to="/"
          style={{
            paddingLeft: "10px",
            color: "inherit",
            textDecoration: "none",
          }}
          onClick={this.logoutSubmit}
        >
          Logout
        </RouterLink>
      </div>
    );
  };

  /**
   * Renders the buttons on the navbar when the user is not authenticated
   */
  renderIsNotLoggedIn = () => {
    return (
      <div>
        <RouterLink
          to="/login"
          style={{ color: "inherit", textDecoration: "none" }}
        >
          Login
        </RouterLink>
        <RouterLink
          to="/register"
          style={{
            paddingLeft: "10px",
            color: "inherit",
            textDecoration: "none",
          }}
        >
          Register
        </RouterLink>
      </div>
    );
  };

  /**
   * Renders the navbar appropriately based on the isLoggedIn prop
   */
  render() {
    return (
      <AppBar position="static">
        <Toolbar>
          <RouterLink
            to="/"
            style={{ color: "inherit", textDecoration: "none" }}
          >
            <Typography variant="inherit">Meowl</Typography>
          </RouterLink>
          <Grid
            container
            alignItems="flex-start"
            justify="flex-end"
            direction="row"
          >
            {this.props.isLoggedIn
              ? this.renderIsLoggedIn()
              : this.renderIsNotLoggedIn()}
          </Grid>
        </Toolbar>
      </AppBar>
    );
  }
}
