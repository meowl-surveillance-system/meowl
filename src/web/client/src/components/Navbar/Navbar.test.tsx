import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { BrowserRouter as Router } from 'react-router-dom';
import Navbar from "./Navbar";

describe("Navbar component", () => {


  it("should render the Navbar component", () => {
  	const { container } = render(<Router><Navbar /></Router>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should call onAuthChange on logoutSubmit", async () => {
  	const onAuthChangeMock = jest.fn();
  	const navbarMock = new Navbar({
  		onAuthChange: onAuthChangeMock
  	});
  	await navbarMock.logoutSubmit();
  	expect(onAuthChangeMock).toHaveBeenCalled();
  });
});