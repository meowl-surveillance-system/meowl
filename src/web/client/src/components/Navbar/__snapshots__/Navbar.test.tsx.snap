// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar component should render the Navbar component 1`] = `
<header
  class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
>
  <div
    class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
  >
    <a
      href="/"
      style="text-decoration: none;"
    >
      <span
        class="MuiTypography-root"
      >
        Meowl
      </span>
    </a>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-flex-start MuiGrid-justify-xs-flex-end"
    >
      <div>
        <a
          href="/login"
          style="text-decoration: none;"
        >
          Login
        </a>
        <a
          href="/register"
          style="padding-left: 10px; text-decoration: none;"
        >
          Register
        </a>
      </div>
    </div>
  </div>
</header>
`;
